feat: ilp-cli tool for interacting with nodes

bstrie opened the pull request:
This PR adds a new command-line tool, `ilp-cli`. It's a beginner-friendly interface for interacting with a running interledger node. The intent is for `ilp-cli` to be for `interledger` what `redis-cli` is for `redis-server`. Right now it only includes as much functionality as is necessary (creating new accounts, posting payments, and getting account balances) to port the simple example from `curl` to `ilp-cli`. Supporting the remainder of the HTTP API will arrive in future PRs.

emschwartz commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+                .long("auth")
+                .env("ILP_CLI_AUTH")
+                .required(true)
+                .help("An authorization key granting access to the designated operation"),
```
This should specify whether this must be for bearer token-based auth

emschwartz commented on Cargo.toml:
```diff
@@ -16,6 +16,7 @@ members = [
   "./crates/interledger-store-redis",
   "./crates/interledger-stream",
   "./crates/interledger-settlement-engines",
+  "./crates/ilp-cli",
```
Not sure about this being called `ilp-cli` rather than `interledger-cli`

bstrie replied:
For the sake of being developer-friendly, I must strenuously object to the idea of requiring users to type that every time they want to issue a command. :P Consider the longest names that you've ever seen from a command-line tool; even "ilp-cli" is stretching the limits of what people will type without grousing. We dearly need a shorter synonym for the project, and if pressed I'd rather rename the `interledger` command to `ilp` or `ilp-node` than rename this to match that one. (Having the crates themselves named extremely long things like `interledger-settlement-ethereum` is less of an issue, because presumably we expect people to `use` them once and never worry again.)

emschwartz replied:
Changed my mind on this. I think it makes sense to call the binaries `ilp-` and the libraries `interledger-`

emschwartz commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+                .env("ILP_CLI_AUTH")
+                .required(true)
+                .help("An authorization key granting access to the designated operation"),
+            Arg::with_name("node_address")
```
In other places we call the URL a client is going to connect to `_url` and the address + port a service is going to bind to `_bind_address`
```suggestion
            Arg::with_name("node_url")
```

emschwartz commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+                .long("node")
+                .env("ILP_CLI_NODE")
+                .default_value("localhost:7770")
+                .help("The URL of the node to which to connect"),
```
```suggestion
                .help("The base URL of the node to connect to"),
```

emschwartz commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+                .env("ILP_CLI_NODE")
+                .default_value("localhost:7770")
+                .help("The URL of the node to which to connect"),
+            Arg::with_name("print_response")
```
As opposed to what?

emschwartz commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+                .help("The URL of the node to which to connect"),
+            Arg::with_name("print_response")
+                .short("p")
+                .long("print-response")
```
I think we have a somewhat inconsistent use of underscores and dashes in our CLIs. Any reason not to use underscores? Mostly thinking about consistency with field names, though I could be convinced otherwise

bstrie replied:
It seems to be general convention in CLIs to use dashes rather than underscores (consider `git cherry-pick`); most CLIs sidestep the issue by never making flags more than a single word. In contrast, Rust identifiers can obviously only contain the latter. In the case here we seek to obviate the need for any manual intervention when converting from CLI arguments into JSON by *exclusively* naming our CLI arguments identically to the JSON fields that the node expects to receive.

I did seek to otherwise impose consistency here: as implemented, the command-line argument names *always* use dashes, and their corresponding entries in the Arguments hashmap will *always* use underscores. The only place that this is ever visible to the user is in the `help` output.

Once thing that I actually have not yet tested is whether Serde is smart enough to automatically deserialize a field named `foo-bar` into `foo_bar`; if that were the case then I would happily use dashes everywhere.

bstrie replied:
I've checked and Serde doesn't accept a JSON field named `foo-bar` when looking to populate a Rust field named `foo_bar` (which is entirely reasonable), so if we're bothered by having dashes for one and underscores for the other we could change all the flags to be underscores instead. However, for now I'm defaulting to leaving it as-is, since it's a low priority style change that can be addressed in the future.

emschwartz commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+        ])
+        .subcommands(vec![
+            // Example: ilp-cli add-account alice ABC 9
+            SubCommand::with_name("add-account")
```
I think we might want to make this more like `ilp-cli accounts add...` because there may be other account-related operations we want to do. What do you think?

gakonst replied:
I agree. 

`interledger-cli accounts <cmd>` where `cmd` matches the [corresponding endpoints](https://github.com/interledger-rs/interledger-rs/blob/master/crates/interledger-api/src/routes/accounts.rs#L388-L400)

```rust
    get_spsp
        .or(get_spsp_well_known)
        .or(post_accounts)
        .or(get_accounts)
        .or(put_account)
        .or(get_account)
        .or(get_account_balance)
        .or(delete_account)
        .or(put_account_settings)
        .or(incoming_payment_notifications)
        .or(post_payments)
        .boxed()
```

bstrie replied:
Absolutely, I'm all for being consistent with the HTTP API here (and not requiring me to think up original names for every subcommand :P ).

emschwartz commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+
+pub fn main() {
+    // Define the arguments to the CLI
+    let mut app = App::new("ilp-cli")
```
Wonder if this should just be `ilp` when you call it

bstrie replied:
We've discussed this and we're going to look at having two binaries, `ilp-cli` and `ilp-node`.

emschwartz commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+                        .required(true)
+                        .help("The scale of the asset associated with this account"),
+                    // Optional, named arguments
+                    Arg::with_name("ilp_address").long("ilp-address").takes_value(true),
```
Each of these should have descriptions about what they are

bstrie replied:
Indeed, rather than having me document what each of these mean only here in the CLI API, there should be a central document that we could refer to. I also don't actually feel qualified to authoritatively claim what many of these arguments even do.

bstrie replied:
Since this is a general issue with our docs I'm leaning towards leaving this as-is for now; I'll put a TODO so that it doesn't just get dropped on the floor.

emschwartz commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+                    .required(true)
+                    .help("The username of the account whose balance to return")),
+            // Example: ilp-cli post-payment alice 500 "http://localhost:8770/accounts/bob/spsp"
+            SubCommand::with_name("post-payment")
```
nit:
```suggestion
            SubCommand::with_name("send-payment")
```

gakonst replied:
That's probably due to how it's in the API https://github.com/interledger-rs/interledger-rs/blob/master/crates/interledger-api/src/routes/accounts.rs#L398

bstrie replied:
If we go with the above suggestion we could have it be `accounts alice payments` instead.

emschwartz commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+                    .help("The username of the account whose balance to return")),
+            // Example: ilp-cli post-payment alice 500 "http://localhost:8770/accounts/bob/spsp"
+            SubCommand::with_name("post-payment")
+                .about("Issue a payment from an account on this node")
```
```suggestion
                .about("Send a payment from an account on this node")
```

emschwartz commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+                    .required(true)
+                    .help("The username of the account whose balance to return")),
+            // Example: ilp-cli post-payment alice 500 "http://localhost:8770/accounts/bob/spsp"
+            SubCommand::with_name("post-payment")
```
This won't actually work with the admin authorization right now. We should either make it so you can send a payment from a user's account using admin auth or make a note that it will only work for the authenticated user

bstrie replied:
Since it seems out of scope to make such a change to the HTTP API right now, I'll leave a TODO to note this as a future improvement.

emschwartz commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+                .about("Issue a payment from an account on this node")
+                .args(&[
+                    Arg::with_name("sender_username")
+                        .index(1)
```
I think positional arguments are slightly worse for something like this because it's hard to remember the order. I think it might be better to stick with named args and just provide the long and short versions of them

gakonst replied:
Yeah I prefer 1 positional argument when there's some subcommand, and then use named args for the args of that subcommand

bstrie replied:
In this case I was using positional arguments for all required arguments and named arguments for all optional arguments, but I'm happy to reconsider.

emschwartz commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+                        .takes_value(true)
+                        .required(true)
+                        // TODO: better way of describing this parameter
+                        .help("The SPSP address of the account receiving the payment"),
```
```suggestion
                        .help("The Payment Pointer or SPSP address of the account receiving the payment (either of the form $bob.wallet.example or http://localhost:7770/accounts/just_an_example/spsp"),
```

emschwartz commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+                "add-account" => {
+                    let args = extract_args(subcommand_matches);
+                    client
+                        // TODO: tacking on the protocol like this doesn't feel ideal,
```
By protocol do you mean `http://` or `/accounts`? The node URL, if specified should include either `http://` or `https://` so we don't have to guess which one it is

emschwartz commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+                    // Final output
+                    Ok(val) => {
+                        if res.status().is_success() {
+                            if matches.is_present("print_response") {
```
Instead of `print_response`, I think it's more common to have an optional `quiet` mode that's off by default

emschwartz commented on crates/interledger-api/src/lib.rs:
```diff
@@ -21,6 +22,48 @@ use secrecy::SecretString;
+
+// Ordinarily providing a serde default for Option types is unnecessary;
+// this is only for cases where we also set the `deserialize_with` attribute.
+pub fn optional_default<T>() -> Option<T> {
```
I believe just using `#[serde(default)]` will have the same effect

emschwartz commented on examples/simple/README.md:
```diff
@@ -214,13 +214,13 @@ Let's create accounts on both nodes. The following script sets up accounts for t
     curl \
         -H "Content-Type: application/json" \
-        -H "Authorization: Bearer admin-a" \
+        -H "Authorization: Bearer admin-example" \
```
Why is this being changed? This is meant to show it using the admin credentials for different nodes

bstrie replied:
I'm under the impression that we consider these examples unrepresentative because we don't actually expect anyone to host two logical nodes on the same machine. Furthermore, for any extensive use of this CLI I would encourage people to use the environment variable to avoid the need to type their password repeatedly. In that sense using different passwords for each logical node seems irrelevant to the example, and requiring that the example include additional command-line flags only to accommodate an atypical use case would seem to be a distraction from what we expect the user to take away from this example.

bstrie replied:
Separate passwords reinstated; now that I've also begin using a symbolic link to get of the `cargo run` invocation, that leaves plenty of room for the occasional `--auth admin-b` argument that the one node will need, and manages to nicely show off the features of the CLI at the same time.

emschwartz commented on examples/simple/README.md:
````diff
@@ -285,109 +285,73 @@ else
 ```bash
 # Insert accounts on Node A
 # One account represents Alice and the other represents Node B's account with Node A
+export ILP_CLI_AUTH=admin-example
````
I think we can either re-export this before specific calls or pass this in as a CLI arg in each call so that we don't need to make all the nodes use the same admin auth token

bstrie replied:
Let's centralize this discussion at https://github.com/interledger-rs/interledger-rs/pull/334#discussion_r328406029.

emschwartz commented on examples/simple/README.md:
```diff
@@ -285,109 +285,73 @@ else
-    "routing_relation": "Peer"}' \
-    http://localhost:7770/accounts
+printf "Creating Alice's account on Node A...\n"
+cargo run --quiet --bin ilp-cli -- \
```
Ugh this command does not look amazing (because of the `cargo run...` part in front). Would it be terrible to do something like aliasing `ilp-cli` to that whole string? If you were really going to install the CLI on your system you'd do `cargo install ilp-cli` so that it would be available in your path without needing `cargo run...`

gakonst replied:
I think we can just add a step in the top level of the examples that they should do `cargo install --release X` and replace all occurrences of `cargo run` with the binary name (or with `./target/{debug,release}/X`

bstrie replied:
> I think we can just add a step in the top level of the examples that they should do cargo install --release X and replace all occurrences of cargo run with the binary name (or with ./target/{debug,release}/X

I'll explore the former, but with regard to the latter we might have trouble using a relative path because we can't guarantee from which directory the user will be running this example; we could require that they be in `examples/simple` if that seems reasonable.

bstrie replied:
In the interest of expedience I'm changing these invocations to just `../../target/debug/ilp-cli`, which could be nicer with an alias or symbolic link, but is a good compromise for the moment and is the simplest and most transparent thing that works.

emschwartz replied:
Maybe it's not so bad. Do we need to run it with `--quiet`?

bstrie replied:
We do want that flag because otherwise you get a few lines of bright green cargo output after every single invocation. I've actually gone even further and begun using a symbolic link here, which ends up looking very nice.

emschwartz commented on examples/simple/README.md:
```diff
@@ -536,10 +501,10 @@ Check out the [other examples](../README.md) for more complex demos that show ot
-        test_equals_or_exit '{"balance":"500"}' test_http_response_body -H "Authorization: Bearer admin-a" http://localhost:7770/accounts/node_b/balance
-        test_equals_or_exit '{"balance":"-500"}' test_http_response_body -H "Authorization: Bearer admin-b" http://localhost:8770/accounts/node_a/balance
-        test_equals_or_exit '{"balance":"500"}' test_http_response_body -H "Authorization: Bearer admin-b" http://localhost:8770/accounts/bob/balance
+        test_equals_or_exit '{"balance":"-500"}' test_http_response_body -H "Authorization: Bearer admin-example" http://localhost:7770/accounts/alice/balance
```
Should the tests also use the CLI? @bstrie @dora-gt

gakonst replied:
Yeah I'd prefer that

dora-gt replied:
hmmmm difficult question.

This simply checks, finally it worked well or not. So using `cli` here means mixing adulterants in. That said, actually now this is working as an integration test, using `cli` makes it possible to test `cli` itself. It might be better.

bstrie replied:
I do like the idea of using this to test the CLI as well; since this is a monorepo I think it's reasonable to make the CI test the examples and the CLI in the same test.

gakonst commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+                .about("Creates a new account on this node")
+                .args(&[
+                    // Required, positional arguments
+                    Arg::with_name("username")
```
I'd prefer if we used named arguments here, what do you think? Combined with the above comment, this would be

`interledger-cli accounts add --username alice --asset-code ABC --asset-scale 9

Also, let's use `-` everywhere in our CLIs instead of underscores?

bstrie replied:
I'll look at implementing the suggestion ( https://github.com/interledger-rs/interledger-rs/pull/334#discussion_r328403702 ), although for consistency with the HTTP API I would probably switch to named arguments for every argument *except* username here, so `interledger-cli accounts add alice --asset-code ABC --asset-scale 9`.

bstrie replied:
> Also, let's use - everywhere in our CLIs instead of underscores?

Let's centralize dicussion of this in https://github.com/interledger-rs/interledger-rs/pull/334#discussion_r328403524 .

gakonst commented on examples/simple/README.md:
```diff
@@ -136,12 +136,12 @@ When you want to watch logs, use the `tail` command. You can use the command lik
         -e ILP_ADDRESS=example.node_a \
         -e ILP_SECRET_SEED=8852500887504328225458511465394229327394647958135038836332350604 \
-        -e ILP_ADMIN_AUTH_TOKEN=admin-a \
+        -e ILP_ADMIN_AUTH_TOKEN=admin-example \
```
we had admin-a and admin-b to show that each node can use a different auth token, think it'd be better to leave that as is

bstrie replied:
I believe that's evident from the `interledger node` command itself. Let's centralize discussion in https://github.com/interledger-rs/interledger-rs/pull/334#discussion_r328406029 .

bstrie commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+        .args(&[
+            Arg::with_name("authorization_key")
+                .long("auth")
+                .env("ILP_CLI_AUTH")
```
> I think it is rather API's auth than CLI's auth, so I feel it a bit strange.
Or... ILP_CLI_NODE_AUTH also makes sense to me.

I would not mind considering other names for this.

> Also, we might want to give the token from STDIN.
> It could be another PR, maybe.

This is implemented already, and indicated by the output of `ilp-cli --help`.

bstrie commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+                    Arg::with_name("ilp_address").long("ilp-address").takes_value(true),
+                    Arg::with_name("max_packet_amount").long("max-packet-amount").takes_value(true),
+                    Arg::with_name("min_balance").long("min-balance").takes_value(true).allow_hyphen_values(true),
+                    Arg::with_name("ilp_over_http_url").long("ilp-over-http-url").takes_value(true),
```
I was just about to submit this PR to your branch when git told me that your branch no longer existed; turns out you had just merged it into master. :P

bstrie commented:
Also note that this initial PR doesn't yet update any of the Docker side of things; wasn't sure if that ought to be left to a future PR.

bstrie commented:
> If we use cli in examples, we have to amend the other examples too.
> I'd be happy to take it on if you want.

@dora-gt, I'm actually using the examples as my to-do list for which APIs to implement first, so no need. :) However, I don't intend on updating the docker side of the examples right now, and I'd be happy to have you work on that (but don't worry about doing that until after I have the examples updated).

bstrie commented:
I've just force-pushed a rebased and squashed version with all changes since the last review.

Things not yet done:

* Most of the CLI subcommands still need proper argument parsing and HTTP request logic. However, if people can confirm that the general direction of the PR looks good, this would be a very straightforward addition given how the code is currently structured.
* This commit also updates the XRP settlement example, but the remaining three examples are not yet updated; I wanted to make sure that the general direction here is correct before proceeding further on that task.

emschwartz commented on Cargo.toml:
```diff
@@ -17,4 +17,5 @@ members = [
   "./crates/interledger-store-redis",
   "./crates/interledger-stream",
   "./crates/interledger-settlement-engines",
+  "./crates/ilp-cli",
```
nit: put this in alphabetical order

emschwartz commented on crates/ilp-cli/src/interpreter.rs:
```diff
@@ -0,0 +1,223 @@
+    // Parse the CLI input using the defined arguments
+    let matches = app.clone().get_matches();
+
+    let node = Node {
```
How about `NodeClient` or something like that? The thing we're interacting with is the "node"

emschwartz commented on crates/ilp-cli/src/parser.rs:
```diff
@@ -0,0 +1,227 @@
+}
+
+fn accounts_create<'a, 'b>() -> App<'a, 'b> {
+    // Example: ilp-cli add-account alice ABC 9
```
This example should be updated to use non-positional arguments

emschwartz commented on crates/ilp-cli/src/parser.rs:
```diff
@@ -0,0 +1,227 @@
+                        .required(true)
+                        .help("The username of the account on this node issuing the payment"),
+                    Arg::with_name("source_amount")
+                        .long("source-amount")
```
What do you think about adding short parameters for this and receiver? (maybe `-a` and `-r`?)

bstrie replied:
I've done this in my branch, part of my original reluctance is that I don't want to end up with the same short flags having different meanings in different subcommands (even though technically this is no problem to implement), so I wanted to ask about which commands to prioritize implementing short flags for (for example, I was considering using `-a` for the `--auth` top-level flag).

emschwartz replied:
That's fair. Maybe not

emschwartz commented on crates/ilp-cli/src/parser.rs:
```diff
@@ -0,0 +1,227 @@
+                        .takes_value(true)
+                        .required(true)
+                        .help("The amount to transfer from the sender to the receiver, denominated in units of the sender's assets"),
+                    Arg::with_name("receiver")
```
Would it be clearer what you need to put here if the parameter were called "payment pointer" instead of "receiver"?

gakonst replied:
As a user, I do not want to know what "payment pointer" means, or have to look at the [RFC](https://interledger.org/rfcs/0026-payment-pointers/). I think "receiver" should be fine here.

emschwartz commented on examples/simple/README.md:
```diff
@@ -201,7 +204,7 @@ printf "\nWaiting for Interledger.rs nodes to start up"
 wait_to_serve "http://localhost:8770" 10 || error_and_exit "\nFailed to spin up nodes. Check out your configuration and log files."
 
-printf "done\nThe Interledger.rs nodes are up and running!\n\n"
+printf " done\nThe Interledger.rs nodes are up and running!\n\n"
```
```suggestion
printf "done\nThe Interledger.rs nodes are up and running!\n\n"
```

bstrie replied:
This space is deliberate because the preceding output is the line of repeated `.` that we print while waiting for the nodes to spin up, so this makes the difference between the output looking like `Waiting for Interledger.rs nodes to start up.....done` and `Waiting for Interledger.rs nodes to start up..... done`.

emschwartz commented on examples/simple/README.md:
````diff
@@ -283,111 +288,72 @@ else
 -->
 
 ```bash
+# We merely use this symbolic link here to pretend that ilp-cli is installed
````
```suggestion
# We merely use this symbolic link here to pretend that ilp-cli is installed
# You would not need this line if you installed `ilp-cli` by running `cargo install ilp-cli`
```

bstrie replied:
My only concern with suggesting that people use `cargo install` is that I worry about a situation where someone has a stale version of the CLI installed which might end up causing errors that are frustrating for users to diagnose down the road. Let me know if you think the risk here isn't worth worrying about.

emschwartz replied:
Not worth it in this case. If the CLI isn't stable right now, it should be very soon

emschwartz commented on examples/simple/README.md:
```diff
@@ -283,111 +288,72 @@ else
+# We merely use this symbolic link here to pretend that ilp-cli is installed
+ln -s ../../target/debug/ilp-cli ilp-cli
+
+# For authenticating to nodes, we can set credentials as follows
```
```suggestion
# For authenticating to nodes, we can set credentials as an environment variable or a CLI argument
```

emschwartz commented on examples/xrp-settlement/README.md:
````diff
@@ -385,77 +389,60 @@ else
 -->
 
 ```bash
+ln -s ../../target/debug/ilp-cli ilp-cli
````
We shouldn't do this if it doesn't work when run from a different directory

bstrie replied:
I'm not sure what the preferred alternative is; this is replacing the hideous `cargo run --quiet --bin ilp-cli --` that otherwise preceded every command and served to obfuscate what the example was trying to show (and is annoyingly slower in practice, since every individual `cargo run` has to check to see if a recompilation should occur). We could `cargo install` perhaps, but that seems like a more heavyweight solution and I fear that we're going to have users accidentally managing to run old versions of the CLI and running into maddening errors.

emschwartz replied:
Replacing the longer command sounds good as long as it works when called from different directories. I don't actually know if that link is relative to the file it's being called in or from the working directory. If it's the working directory, is there a way to force it to be from the perspective of the file?

emschwartz commented on examples/xrp-settlement/README.md:
````diff
@@ -385,77 +389,60 @@ else
 
 ```bash
+ln -s ../../target/debug/ilp-cli ilp-cli
+export ILP_CLI_API_AUTH=hi_alice
````
If this is only used in once, I think it would be better to pass it in as a CLI arg

bstrie replied:
The env variable gets used in six of the thirteen `ilp-cli` invocations in this example (and could be used in an additional one if admin auth could be used to send payments, as you had mentioned was something we were looking at implementing). In lieu of having an alternative way to authenticate (e.g. a `sudo`-like setup, or a file containing credentials, or an interactive session requiring the user to type their password every time) I'd suggest that the environment variable should be the recommended way of providing credentials, with the stdin approach providing a fallback.

emschwartz commented on examples/xrp-settlement/README.md:
```diff
@@ -385,77 +389,60 @@ else
+    --settle-threshold 500 \
+    --min-balance -1000 \
+    --settle-to 0 \
+    --routing-relation Peer > logs/account-bob-alice.log &
```
Did you mean to get rid of `2>/dev/null`?

bstrie replied:
Sort of, I wasn't sure why we were silencing stderr output and figured that I would err on the side of being too verbose rather than accidentally quelching an error that we'd rather notice.

gakonst commented on crates/ilp-cli/src/interpreter.rs:
```diff
@@ -0,0 +1,223 @@
+                        exit(1);
+                    }
+                },
+                "status" => node.get_root(ilp_cli_matches),
```
There's a few commands currently which do not require any authorization, and as such the `--auth` flag should not be required. These are: `/status`, GET `/rates`, GET `/routes`. @emschwartz not sure if you'd want to add auth on the rates and routes endpoints, probably not.

bstrie replied:
If this is the case then the most straightforward solution seems to be to move the `--auth` flag to exist only underneath those subcommands that require it, since this "uppermost" level is intended for arguments that apply to all subcommands.

gakonst commented on crates/ilp-cli/src/interpreter.rs:
```diff
@@ -0,0 +1,223 @@
+                        }
+                        command => panic!("Unhandled `ilp-cli accounts` subcommand: {}", command),
+                    },
+                    // TODO: subcommand-specific help output
```
Yeah that'd be helpful. Also, when i run `./ilp-cli`, I'd prefer if I got the full output of `./ilp-cli --help` (as below) instead of being prompted to add the `--help` flag for more info

```
./target/debug/ilp-cli --help
ilp-cli 0.0.1
Interledger.rs Command-Line Interface

USAGE:
    ilp-cli [FLAGS] [OPTIONS] --auth <authorization_key> [SUBCOMMAND]

FLAGS:
    -h, --help       Prints help information
    -q, --quiet      Disable printing the bodies of successful HTTP responses upon receipt
    -V, --version    Prints version information

OPTIONS:
        --auth <authorization_key>    An HTTP bearer authorization token permitting access to the designated operation
                                      [env: ILP_CLI_API_AUTH=]
        --node <node_url>             The base URL of the node to connect to [env: ILP_CLI_NODE_URL=]  [default:
                                      http://localhost:7770]

SUBCOMMANDS:
    accounts              Operations for interacting with accounts
    help                  Prints this message or the help of the given subcommand(s)
    pay                   Send a payment from an account on this node
    rates
    routes
    settlement-engines
    status
```

gakonst commented on crates/ilp-cli/src/interpreter.rs:
```diff
@@ -0,0 +1,223 @@
+                        exit(1);
+                    }
+                },
+                "pay" => node.post_account_payments(ilp_cli_matches),
```
How about making the sender provided as `--from` and rename all `--receiver` instances to `--to` and `--source-amount` to `--amount` or `--value`? Think this is easier to type. I understand that `source-amount` is more accurate since the exactly the same amount will not get sent to the receiver, but once we get quoting working we could perhaps make the functionality match the looser language. 

```
./target/debug/ilp-cli pay --help
ilp-cli-pay
Send a payment from an account on this node

USAGE:
    ilp-cli --auth <authorization_key> pay <sender_username> --receiver <receiver> --source-amount <source_amount>

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

OPTIONS:
        --receiver <receiver>              The Payment Pointer or SPSP address of the account receiving the payment
        --source-amount <source_amount>    The amount to transfer from the sender to the receiver, denominated in units
                                           of the sender's assets

ARGS:
```

bstrie replied:
My default stance was to name the command line arguments identically to the field names that we expect in the JSON payload (in case somebody wanted to go from using this tool to using the HTTP API directly), but I'm happy to make this more ergonomic, and in fact I can even implement aliases for these arguments so that people can type either `--to` or `--receiver` at their discretion.

gakonst commented on crates/ilp-cli/src/interpreter.rs:
```diff
@@ -21,203 +21,252 @@ pub fn run<'a, 'b>(mut app: App<'a, 'b>) {
-    fn put_rates(&self, matches: &ArgMatches) -> reqwest::Result<Response> {
-        unimplemented!()
+    fn put_rates(&self, matches: &ArgMatches) -> Result<Response, Error> {
+        let auth = matches.value_of("authorization_key").unwrap();
```
If this fails will it show that the authorization key was not found? Maybe replace the unwrap with a more verbose expect message?

bstrie replied:
I should sprinkle some more comments around here, basically the gist of this file is that anywhere you see an `unwrap` it's "impossible" to actually trigger a panic because if the field were omitted then Clap would have already errored out during the `get_matches` stage (I say "impossible" in quotes because there could always be a bug in Clap or something).

bstrie replied:
To be clear here's what the message looks like today if this field is omitted:
```
⌁ ./ilp-cli rates set-all -r a 1.0
error: The following required arguments were not provided:
    --auth <authorization_key>

USAGE:
    ilp-cli rates set-all --auth <authorization_key> --rate <rate> <rate>

For more information try --help
```

gakonst commented on crates/ilp-cli/src/parser.rs:
```diff
@@ -0,0 +1,365 @@
+}
+
+fn routes_set<'a, 'b>() -> App<'a, 'b> {
+    SubCommand::with_name("set")
```
```suggestion
    SubCommand::with_name("set").about("Set the routes for this node")
```

gakonst commented on crates/ilp-cli/src/main.rs:
```diff
@@ -0,0 +1,178 @@
+}
+
+#[cfg(test)]
+// Note that this module contains interface tests, not integration tests.
```
Great idea.

emschwartz commented on crates/ilp-cli/src/parser.rs:
```diff
@@ -0,0 +1,365 @@
+    SubCommand::with_name("accounts")
+        .about("Operations for interacting with accounts")
+        .arg(
+            Arg::with_name("is_admin")
```
If this is only used for the update settings method, I think we should just have two different method names for those operations. In general, the node will figure out whether you're doing stuff as the admin or not

emschwartz commented on crates/ilp-cli/src/parser.rs:
```diff
@@ -0,0 +1,365 @@
+    SubCommand::with_name("balance")
+        .about("Returns the balance of an account")
+        .args(&[
+            Arg::with_name("authorization_key")
```
nit: this is repeated a lot, what about putting it in a lazy_static or separate function?

emschwartz commented on crates/ilp-cli/src/parser.rs:
```diff
@@ -0,0 +1,365 @@
+                .env("ILP_CLI_API_AUTH")
+                .required(true)
+                .help("An HTTP bearer authorization token permitting access to this operation"),
+            Arg::with_name("asset_code")
```
Since these fields are repeated in two methods, let's put them in a separate function so we only have to update the fields in one place

bstrie replied:
I was thinking about that and it got me thinking about whether or not these two sets of arguments ought to be slightly different; it might make sense to make `asset_code` and `asset_scale` optional for updating an account, while keeping it mandatory for creating an account. But in lieu of that I'll break this out into its own thing for now.

emschwartz commented on examples/simple/README.md:
```diff
@@ -283,111 +288,74 @@ else
-    "routing_relation": "Peer"}' \
-    http://localhost:7770/accounts
+printf "Creating Alice's account on Node A...\n"
+./ilp-cli --quiet accounts create alice \
```
Beautiful

emschwartz commented on examples/xrp-settlement/README.md:
```diff
@@ -385,77 +389,62 @@ else
+    --settle-threshold 500 \
+    --min-balance -1000 \
+    --settle-to 0 \
+    --routing-relation Peer > logs/account-alice-bob.log &
```
We should make sure it doesn't care about the capitalization of this

bstrie replied:
Currently `ilp-cli` deliberately does as little inspection and processing of the arguments as possible, so if the user passes "peer" or "Peer" or "pEeR" it will just forward it as-is to the server; the general mindset here is that we could focus on improving error messages on the server side rather than duplicating validation on both client and server. If we did want validation we could certainly use https://docs.rs/clap/2.33.0/clap/struct.Arg.html#method.possible_values combined with https://docs.rs/clap/2.33.0/clap/struct.Arg.html#method.case_insensitive to make it work, but I'll call that low-priority for now (and if we're going to make any TODOs here, it should be to hash out a policy regarding client-side vs. server-side validation).
